crypto/elliptic-curve.ts

```diff
@@ -703,3 +716,369 @@ function createCurveAffine({
+
+  // B = (X + Y)^2
+  let B = mod((X + Y) ** 2n, p);
+  // C = X^2
+  let C = mod(X * X, p);
+  // D = Y^2
+  let D = mod(Y * Y, p);
+  // E = a * C
+  let E = mod(a * C, p);
+  // F = E + D
+  let F = mod(E + D, p);
+  // H = Z^2
+  let H = mod(Z * Z, p);
+  // J =  F - 2 * H
+  let J = mod(F - 2n * H, p);
+
+  return {
+    x: mod((B - C - D) * J, p),
+    y: mod(F * (E - D), p),
+    z: mod(F * J, p),
+  };
+}
+
+function projectiveScaleTwisted(
+  g: GroupProjective,
+  x: bigint | boolean[],
+  p: bigint,
+  a: bigint,
+  d: bigint
+) {
+  let bits = typeof x === 'bigint' ? bigIntToBits(x) : x;
+  let h = projectiveZeroTwisted;
+  for (let bit of bits) {
+    if (bit) h = projectiveAddTwisted(h, g, p, a, d);
+    g = projectiveDoubleTwisted(g, p, a);
+  }
+  return h;
+}
+
+function projectiveFromTwisted({
+  x,
+  y,
+  infinity,
+}: GroupTwisted): GroupProjective {
+  if (infinity) return projectiveZeroTwisted;
+  return { x, y, z: 1n };
+}
+
+// The twisted curve with equation
+// a * x^2 + y^2 = 1 + d * x^2 * y^2
+// in projective coordinates is represented as
+// a * X^2 * Z^2 + Y^2 Z^2 = Z^4 + d * X^2 * Y^2
+// where x = X/Z, y = Y/Z, and Z ≠ 0
+function projectiveToTwisted(g: GroupProjective, p: bigint): GroupTwisted {
+  let z = g.z;
+  if (z === 0n) {
+    // degenerate case
+    return twistedZero;
+  } else if (z === 1n && g.x === 0n && g.y === 1n) {
+    // special case for the zero point
+    return twistedZero;
+  } else if (z === 1n) {
+    // any other normalized affine form
+    return { x: g.x, y: g.y, infinity: false };
+  } else {
+    let zinv = inverse(z, p)!; // we checked for z === 0, so inverse exists
+    // x/z
+    let x = mod(g.x * zinv, p);
+    // y/z
+    let y = mod(g.y * zinv, p);
+    return { x: x, y: y, infinity: false };
+  }
+}
+
+type CurveTwisted = ReturnType<typeof createCurveTwisted>;
+
+// Creates twisted Edwards curves in the form
+// a * x^2 + y^2 = 1 + d * x^2 * y^2
+// with a ≠ 0, d ≠ 0 and a ≠ d
+function createCurveTwisted({
+  name,
+  modulus: p,
+  order,
+  cofactor,
+  generator,
+  a,
+  d,
+}: TwistedCurveParams) {
+  let hasCofactor = cofactor !== undefined && cofactor !== 1n;
+
+  const Field = createField(p);
+  const Scalar = createField(order);
+  const one = { ...generator, infinity: false };
+  const Endo = undefined; // for Ed25519
+
+  assert(a !== 0n, 'a must not be zero');
+  assert(d !== 0n, 'd must not be zero');
+  assert(a !== d, 'a must not be equal to d');
+
+  return {
+    name,
+    /**
+     * Arithmetic over the base field
+     */
+    Field,
+    /**
+     * Arithmetic over the scalar field
+     */
+    Scalar,
+
+    modulus: p,
+    order,
+    a,
+    d,
+    cofactor,
+    hasCofactor,
+
+    zero: twistedZero,
+    one,
+
+    hasEndomorphism: Endo !== undefined,
+    get Endo() {
+      if (Endo === undefined) throw Error(`no endomorphism defined on ${name}`);
+      return Endo;
+    },
+
+    from(g: { x: bigint; y: bigint }): GroupTwisted {
```

Comment:
Don't you want to enforce to have x and y given as canonical values? (i.e. smaller than the order of the base field?). I would be highly in favor of this.

Reply:
I guess no, and then perform the operations modulo the order and check that it holds (like, {x: 0, y: 1+p} is the identity instead of failing). But I am open to other alternatives, whatever is the standard.